GitHub Actions

srivarra opened the pull request:
**If you haven't already, please read through our [contributing guidelines](https://ark-analysis.readthedocs.io/en/latest/_rtd/contributing.html) before opening your PR**

**What is the purpose of this PR?**

Adds GitHub Actions for building and testing TMI.

**How did you implement your changes**

Added the `build.yml` in `workflows` and added CIBuildWheel to `pyproject.toml`.

Builds *Pure Python* Wheels, which can run on any version of Python 3, independent of the Operating System and underlying hardware. Caches the majority of requirements as well, and uploads the coverage to Coveralls.

**Remaining issues**

- [x] So it looks we need to change the name of TMI as it's already taken on PyPI
- [x] Get TestPyPI to work

srivarra commented:
@ngreenwald Will make a new release after this is merged.

ngreenwald commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,124 @@
+          packages_dir: dist/
+          verbose: true
+
+  pypi_publish:
```
What's the difference between these two? What workflow are you imagining for testing and then running the actual pypi?

srivarra replied:
For every PR, the `build_wheels_sdist` job will always run tests, upload coverage, and build wheels.

For every new release, the workflow will run `build_wheels_sdist` and then if and only if the `build_wheels_sdist` job succeeds, then `test_pypi_publish` will run. Once `test_pypi_publish` runs to succession, it will then upload the wheels to pypi in `pypi_publish`.

ngreenwald reviewed (changes requested):
Just a question about the test pypi/pypi workflow

alex-l-kong reviewed (commented):
Nothing to add on top of Noah's comment. I'm really loving Alpineer!

srivarra commented:
@alex-l-kong @ngreenwald @camisowers Would you guys prefer to split the file apart, one for running tests, and the other for running pypi publishing, or keep it as one monolithic workflow?

camisowers commented:
Looks really clean Sri! Excited for a tests folder in the other repos too.

alex-l-kong commented:
> @alex-l-kong @ngreenwald @camisowers Would you guys prefer to split the file apart, one for running tests, and the other for running pypi publishing, or keep it as one monolithic workflow?

Where possible I think splitting would make it easier to read, but if it's too much work to configure then just leave it as a whole file.

ngreenwald commented:
My thinking was that this current workflow requires a new release in order to test pypi deployment. This means that if testpypi fails after generating a new release, we'd need a subsequent release to fix any errors, right?

In my mind, it would be better to have a way to run test_pypi in the pre-release PR that changes the release name in the various settings file, similar to how you guys would check if the docker built succesfully by toggling the `on_push` tag in Travis. 

Does that make sense, or am I not understanding what the workflow would look like?

srivarra commented:
@ngreenwald 

> My thinking was that this current workflow requires a new release in order to test pypi deployment. This means that if testpypi fails after generating a new release, we'd need a subsequent release to fix any errors, right?

Yes that would be correct. I've made the following changes:
- Split apart the GHA into _ci_, _pypi_publish_, and _test_
- Implemented a similar system when we did the `on_push`, where on a release branch labeled `releases/**` it will run wheel building + testing jobs, as well as the Test PyPI publish job. So we don't need to mess with tags, as long as the branch is named `releases/**` it will run the Test PyPI job.
- In addition I set the `ci` and the `pypi_publish` workflows as `workflow_dispatch` so when this gets merged, we can manually press a button and let the `test_pypi_publish` job run as well.

srivarra commented:
@ngreenwald Are there any other adjustments we'd like to make? If not, this is good to go!
